Kernel/FileSystem/VirtualFileSystem.cpp

```diff
@@ -129,9 +153,25 @@ ErrorOr<void> VirtualFileSystem::unmount(Custody& mountpoint_custody)
             auto mountpoint_path = TRY(mount->absolute_path());
             if (custody_path->view() != mountpoint_path->view())
                 continue;
-            TRY(mount->guest_fs().prepare_to_unmount());
-            mount->guest_fs().decrement_mounted_count({});
-            dbgln("VirtualFileSystem: Unmounting file system {}...", mount->guest_fs().fsid());
+            NonnullRefPtr<FileSystem> fs = mount->guest_fs();
+            TRY(fs->prepare_to_unmount());
+            fs->mounted_count({}).with([&](auto const& count) {
+                if (count == 1) {
+                    dbgln("VirtualFileSystem: Unmounting file system {} for the last time...", fs->fsid());
+                    m_file_systems_list.with([&](auto& list) {
+                        list.remove(*fs);
+                    });
+                    if (fs->is_file_backed()) {
+                        dbgln("VirtualFileSystem: Unmounting file backed file system {} for the last time...", fs->fsid());
+                        auto& file_backed_fs = static_cast<FileBackedFileSystem&>(*fs);
+                        m_file_backed_file_systems_list.with([&](auto& list) {
+                            list.remove(file_backed_fs);
+                        });
+                    }
+                }
+            });
+            fs->decrement_mounted_count({});
```

Comment:
This looks like it will lock the filesystem mount count twice in a row, and smells like TOCTOU.

Can we make decrement_mounted_count take a callback for `on_last_count` or `on_will_be_really_unmounted` or similar?